pico_sdk_init.cmake

```diff
@@ -1,54 +1,64 @@
-# Initialize the Raspberry Pi Pico SDK
+# Pre-initialize the Raspberry Pi Pico SDK, setting up the platform and toolchain and some CMake utility functions
 # This file must be included prior to the project() call
 
-if (_PICO_SDK_INIT)
-    return()
-endif ()
-set(_PICO_SDK_INIT 1)
+# Note: this file is perhaps named badly, as it provides a method pico_sdk_init which
+# the enclosing project calls LATER to actually "initialize" the SDK (by including the CMakeLists.txt from this
+# same directory)
 
-function(pico_is_top_level_project VAR)
-    string(TOLOWER ${CMAKE_CURRENT_LIST_DIR} __list_dir)
-    string(TOLOWER ${CMAKE_SOURCE_DIR} __source_dir)
-    if (__source_dir STREQUAL __list_dir)
-        set(${VAR} 1 PARENT_SCOPE)
-    else()
-        set(${VAR} 0 PARENT_SCOPE)
-    endif()
-endfunction()
+if (NOT TARGET _pico_sdk_pre_init_marker)
+    add_library(_pico_sdk_pre_init_marker INTERFACE)
 
-if (NOT PICO_SDK_PATH)
-    set(PICO_SDK_PATH ${CMAKE_CURRENT_LIST_DIR})
-endif ()
+    function(pico_is_top_level_project VAR)
+        string(TOLOWER ${CMAKE_CURRENT_LIST_DIR} __list_dir)
+        string(TOLOWER ${CMAKE_SOURCE_DIR} __source_dir)
+        if (__source_dir STREQUAL __list_dir)
+            set(${VAR} 1 PARENT_SCOPE)
+        else()
+            set(${VAR} 0 PARENT_SCOPE)
+        endif()
+    endfunction()
 
-get_filename_component(PICO_SDK_PATH "${PICO_SDK_PATH}" REALPATH BASE_DIR "${CMAKE_BINARY_DIR}")
+    if (NOT PICO_SDK_PATH)
+        set(PICO_SDK_PATH ${CMAKE_CURRENT_LIST_DIR})
+    endif ()
 
-set(PICO_SDK_PATH ${CMAKE_CURRENT_LIST_DIR} CACHE PATH "Path to the Raspberry Pi Pico SDK" FORCE)
+    get_filename_component(PICO_SDK_PATH "${PICO_SDK_PATH}" REALPATH BASE_DIR "${CMAKE_BINARY_DIR}")
 
-list(APPEND CMAKE_MODULE_PATH ${PICO_SDK_PATH}/cmake)
+    set(PICO_SDK_PATH ${CMAKE_CURRENT_LIST_DIR} CACHE PATH "Path to the Raspberry Pi Pico SDK" FORCE)
 
-include(${CMAKE_CURRENT_LIST_DIR}/pico_sdk_version.cmake)
-include(pico_utils)
+    list(APPEND CMAKE_MODULE_PATH ${PICO_SDK_PATH}/cmake)
 
-message("PICO_SDK_PATH is ${CMAKE_CURRENT_LIST_DIR}")
+    include(${CMAKE_CURRENT_LIST_DIR}/pico_sdk_version.cmake)
+    include(pico_utils)
 
-include(pico_pre_load_platform)
+    message("PICO_SDK_PATH is ${CMAKE_CURRENT_LIST_DIR}")
 
-# todo perhaps this should be included by the platform instead?
-# We want to configure correct toolchain prior to project load
-include(pico_pre_load_toolchain)
+    include(pico_pre_load_platform)
 
-macro(pico_sdk_init)
-    if (NOT CMAKE_PROJECT_NAME)
-        message(WARNING "pico_sdk_init() should be called after the project is created (and languages added)")
-    endif()
-    add_subdirectory(${PICO_SDK_PATH} pico-sdk)
-endmacro()
+    # todo perhaps this should be included by the platform instead?
```

Comment:
It's a shame that the GitHub web-diff-viewer gets so badly confused by a change-of-indent :confused:

Reply:
there is a "hide whitespace changes" under the gear icon (which you can flip temporarily)

Reply:
Ahh, that's super handy, thanks! :heart: